### PR TITLE
Improve `save_model` and `load_model`

### LINKED ISSUE #19793: model.keras format much slower to load
Anyone experiencing unreasonably slow load times when loading a keras-format saved model? I have noticed this repeated when working in ipython, where simply instantiating a model via `Model.from_config` then calling `model.load_weights` is much (several factors) faster than loading a `model.keras` file.

My understanding is the keras format is simply a zip file with the config.json file and weights h5 (iirc) but weirdly enough, there's something not right going on while loading.

### PR DESCRIPTION
**Pull request by james77777778**
Fixes #19793

- `save_model` will first open the .h5 (in the same working dir as `filepath`), then write it to the zipfile at the end.
- `load_model` will first extract the .h5, then load it using `h5py.File`. The weights file will be deleted at the end.

Result:
Using GPT2 from KerasNLP
|Branch|Function|Memory Usage|Time Cost|
|-|-|-|-|
|Master|`save_model`|1860.43MiB|**3.84s**|
|PR|`save_model`|**1384.17MiB (-25.6%)**|4.48s (+16.7%)|
|||||
|Master|`load_model`|1377.24MiB|26.93s|
|PR|`load_model`|**1376.43MiB (-0.1%)**|**2.10s (-92.3%)**|

The only regression is that `save_model` might be slightly slower than before.

### CODE REVIEW

**Comment by Grvzard** [comment]
a tiny suggestion: the form `_VARS_FNAME + ".h5"` appears many times. Would it be better to define a new constant like `_VARS_FNAME_H5 = _VARS_FNAME + ".h5"`?

**Comment by james77777778** [comment]
> a tiny suggestion: the form `_VARS_FNAME + ".h5"` appears many times. Would it be better to define a new constant like `_VARS_FNAME_H5 = _VARS_FNAME + ".h5"`?

That's cleaner. Thanks for the suggestion.
Updated.

File: keras/src/saving/saving_lib.py [hunk 1]
```diff
@@ -202,6 +227,8 @@ def _load_model_from_fileobj(fileobj, custom_objects, compile, safe_mode):
         weights_store.close()
         if asset_store:
             asset_store.close()
+        if weights_file_path:
```
Comment (fchollet) [inline keras/src/saving/saving_lib.py:230]: We should put the `unlink` in a try/except/finally block to make sure it gets executed no matter what.

File: keras/src/saving/saving_lib.py [hunk 2]
```diff
@@ -133,6 +145,8 @@ def _save_model_to_fileobj(model, fileobj, weights_format):
         )
         weights_store.close()
         asset_store.close()
+        if weights_file_path:
+            zf.write(weights_file_path, weights_file_path.name)
```
Comment (fchollet) [inline keras/src/saving/saving_lib.py:149]: Same here

**Review by fchollet — COMMENTED** [review]
Thanks for the fix! This is great.

File: keras/src/saving/saving_lib.py (hunk 1 above)
Comment (james77777778) [reply keras/src/saving/saving_lib.py]: I have added a try/finally block.
In `_save_model_to_fileobj`, if an error is raised, the bad `model.weights.h5` won't be written to the zip file.
In `_load_model_from_fileobj`, if an error is raised, the extracted `model.weights.h5` will be deleted no matter what.

Tests have been added to verify these behaviors.

**Review by fchollet — APPROVED** [review]
Thank you for the contribution!

**Pull request merged**